service/raft/group0_state_machine.cc

```diff
@@ -374,9 +383,23 @@ void group0_state_machine::drop_snapshot(raft::snapshot_id id) {
 }
 
 future<> group0_state_machine::load_snapshot(raft::snapshot_id id) {
-    // topology_state_load applies persisted state machine state into
-    // memory and thus needs to be protected with apply mutex
     auto read_apply_mutex_holder = co_await _client.hold_read_apply_mutex(_abort_source);
+    if (_in_memory_state_machine_enabled) {
+        co_await reload_state();
+    }
+}
+
+future<> group0_state_machine::enable_in_memory_state_machine() {
+    auto read_apply_mutex_holder = co_await _client.hold_read_apply_mutex(_abort_source);
+    if (!_in_memory_state_machine_enabled) {
+        _in_memory_state_machine_enabled  = true;
+        co_await reload_state();
+    }
+}
+
+future<> group0_state_machine::reload_state() {
+    // we assume that the apply mutex is held, topology_state_load applies
+    // persisted state machine into memory so it needs to be protected with it
```

Comment:
Isn't the header file a better place for such a comment?

Reply:
Perhaps it is, but I just kept the existing comment. I didn't touch this comment at all, I reused most of the body of the `load_snapshot` function and kept the comment, it's just that the diff algorithm chose to show it as a removal and addition.

I agree with about headers being a better place for preconditions (preconditions are a part of interface), but I'd rather leave it as is and minimize the amount of changes. It's a private method.